stacks-signer/src/signer.rs

```diff
@@ -0,0 +1,2005 @@
+            .get_account_nonce(signer_address)
+            .map_err(|e| {
+                warn!(
+                    "Signer #{}: Failed to get account nonce for signer: {e:?}",
+                    self.signer_id
+                );
+            })
+            .unwrap_or(0);
+
+        let current_transactions = self.get_signer_transactions(stacks_client, current_reward_cycle).map_err(|e| {
+            warn!("Signer #{}: Failed to get old transactions: {e:?}. Defaulting to account nonce.", self.signer_id);
+        }).unwrap_or_default();
+
+        for transaction in current_transactions {
+            let origin_nonce = transaction.get_origin_nonce();
+            let origin_address = transaction.origin_address();
+            if origin_address == *signer_address && origin_nonce >= next_nonce {
+                next_nonce = origin_nonce.wrapping_add(1);
+            }
+        }
+        next_nonce
+    }
+
+    /// broadcast the dkg vote transaction according to the current epoch
+    fn broadcast_dkg_vote(
+        &mut self,
+        stacks_client: &StacksClient,
+        new_transaction: StacksTransaction,
+        epoch: StacksEpochId,
+        current_reward_cycle: u64,
+    ) -> Result<(), ClientError> {
+        let txid = new_transaction.txid();
+        let aggregate_key = retry_with_exponential_backoff(|| {
+            stacks_client
+                .get_approved_aggregate_key(self.reward_cycle)
+                .map_err(backoff::Error::transient)
+        })
+        .unwrap_or(None);
+        match epoch {
+            StacksEpochId::Epoch25 => {
+                debug!("Signer #{}: Received a DKG result while in epoch 2.5. Broadcast the transaction to the mempool.", self.signer_id);
+                if aggregate_key.is_none() {
+                    stacks_client.submit_transaction(&new_transaction)?;
+                    info!(
+                        "Signer #{}: Submitted DKG vote transaction ({txid:?}) to the mempool",
+                        self.signer_id
+                    )
+                } else {
+                    debug!("Signer #{}: Already have an aggregate key for reward cycle {}. Do not broadcast the transaction ({txid:?}).", self.signer_id, self.reward_cycle);
+                }
+            }
+            StacksEpochId::Epoch30 => {
+                debug!("Signer #{}: Received a DKG result while in epoch 3.0. Broadcast the transaction only to stackerDB.", self.signer_id);
+            }
+            _ => {
+                debug!("Signer #{}: Received a DKG result, but are in an unsupported epoch. Do not broadcast the transaction ({}).", self.signer_id, new_transaction.txid());
+                return Ok(());
+            }
+        }
+        // For all Pox-4 epochs onwards, broadcast the results also to stackerDB for other signers/miners to observe
+        // TODO: if we store transactions on the side, should we use them rather than directly querying the stacker db slot?
```

Comment:
I'm not sure it's necessary to store them on the side if the stackerdb has a copy already.

Reply:
I will remove the TODO for now. We can always look into the need for this based on performance.